swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -153,6 +153,13 @@ private module Cached {
     or
     nodeFrom.asExpr() = nodeTo.asExpr().(OptionalEvaluationExpr).getSubExpr()
     or
+    // flow through optional binding `if let`
+    exists(ConditionElement ce, ConcreteVarDecl v |
+      ce.getInitializer() = nodeFrom.asExpr() and
+      ce.getPattern() = v.getParentPattern() and
+      nodeTo.asDefinition().getSourceVariable() = v
+    )
+    or
```

Comment:
Does this mean we have flow from `source()` to both `x` and `y` in something like:
```swift
func source() -> (Int, Int)? { ... }

if let (x, y) = source() {
  print(x, y)
}
```

I would have hoped that this case wasn't needed, but that it would be covered by the very first disjunct in this predicate (possibly by extending `assigns`):
```ql
def.(Ssa::WriteDefinition).assigns(nodeFrom.getCfgNode()) and
nodeTo.asDefinition() = def
```

Reply:
I've added your example to the tests.

We do have flow from `source()` to both `x` and `y` here, provided by the `if let` case in `DataFlowPrivate.qll`.  Its needed for all the new results (except line 303) I think.  The SSA assignment case doesn't cover them, but perhaps it could???

Reply:
Sorry if I wasn't clear. What I'm saying is two-fold:
1. I would not expect dataflow from `source()` to `x` and `y` as _dataflow_ step. If anything, it should be a _taint_ step.
2. I don't think the case added should be here. Instead, it should be part of the [first case in this predicate](https://github.com/github/codeql/blob/18a815ca8b5fb01d8867b188a4e5213e7d530154/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L119), where `assigns` is extended to transfer flow from `a` to `b` when it's of the form `if let a = b`.

Does that make sense?

Reply:
> I would not expect dataflow from source() to x and y as dataflow step. If anything, it should be a taint step.

Good point.  In general I've been happy that these are dataflow steps, but this case is an exception.

I'll take a look at your proposed fix (when I can find 15 minutes)...

Reply:
... I've updated the logic to use `Ssa::Definition` in place of `ConcreteVarDecl`, and then updated it again moving that into `Ssa::WriteDefinition.assigns` itself.  That feels like probably an improvement but I'm not very comfortable in the SSA library internals, and I'm not sure what else this change now affects.  I would appreciate your thoughts.

Regarding the `if let (x, y) = source()` case in particular, I think a full fix might involve decoding the `Pattern` on the left side and matching it against access paths of the thing on the right side of the assignment ... potentially a bit of a huge time sink.  I'm inclined to just exclude cases where more than one variable is initialized by the `ConditionElement`.  Or write a taint flow case for these cases perhaps.  Do you have any better ideas?

Reply:
> That feels like probably an improvement but I'm not very comfortable in the SSA library internals, and I'm not sure what else this change now affects. I would appreciate your thoughts.

I think what you've done in 2c35af51cd694a3456f6b8de4283b5f899df32da makes a lot of sense. IIRC, the `assigns` predicate is only used in `localFlowStepCommon`. So your change is still a very local one 👍.

> Regarding the if let (x, y) = source() case in particular, I think a full fix might involve decoding the Pattern on the left side and matching it against access paths of the thing on the right side of the assignment ... potentially a bit of a huge time sink.

Yeah, I completely agree. This proposed pattern decoding should already be covered by https://github.com/github/codeql/pull/11111, no? Of course, without a real implementation of `source()` that actually produces a tuple containing a tainted element, field flow won't kick in in the testcase you've added in https://github.com/github/codeql/pull/11836/commits/4e5483744f9124584dce02b346f847477c83a604 (which is totally fine!).

> I'm inclined to just exclude cases where more than one variable is initialized by the ConditionElement. Or write a taint flow case for these cases perhaps. Do you have any better ideas?

I think this is exactly the right approach, yes. We should have dataflow in the case where just one variable is initialized, but as soon as it's pattern that's producing more than one variable we shouldn't have dataflow.

I haven't fully grasped whether we want taintflow in this case. It will give us taint FPs on something like:
```swift
func second_is_tainted() -> (Int, Int)? {
  return (0, source())
}

...
if let (x, y) = second_is_tainted() {
  sink(x)
}
```

but, of course, it might actually give us some good taintflow in case a query is looking for something like:
```swift
func data_from_network() -> (Int, Int)? {
  string from_network = receive_data()
  return __internal_decode_data(from_network) // let's say we don't have the body of this function.
}

...
if let (x, y) = data_from_network() {
  sink(x)
}
```

Reply:
> This proposed pattern decoding should already be covered by #11111, no? Of course, without a real implementation of source() that actually produces a tuple containing a tainted element, field flow won't kick in in the testcase you've added

I see what you mean about `source()` in this test case.  I'm not sure if content flow through tuples will need a special case for decoding `if let (x, y)` as well.  I can't imagine any of the existing cases quite cover it successfully (I could be wrong).  And your two examples above show that we will need to consider a variety of cases quite carefully when we do get around to solving these sorts of problems.

> We should have dataflow in the case where just one variable is initialized, but as soon as it's pattern that's producing more than one variable we shouldn't have dataflow.

I've added an exclusion (and a commit with a QL-for-QL fix after that).  I'm not convinced its the most accurate / concise way of expressing it so let me know if you can think of a better way.